src/msw/textctrl.cpp

```diff
@@ -2720,12 +2728,50 @@ wxMenu *wxTextCtrl::MSWCreateContextMenu()
 
 void wxTextCtrl::MSWUpdateFontOnDPIChange(const wxSize& newDPI)
 {
-    // Don't do anything for the rich edit controls, they (somehow?) update
-    // their appearance on their own and changing their HFONT, as the base
-    // class version does, would reset all the styles used by them when the DPI
-    // changes, which is unwanted.
+    // Don't use MSWUpdateFontOnDPIChange for the rich edit controls, they
+    // (somehow?) update their appearance on their own and changing their
+    // HFONT, as the base class version does, would reset all the styles used
+    // by them when the DPI changes, which is unwanted.
     if ( !IsRich() )
+    {
         wxTextCtrlBase::MSWUpdateFontOnDPIChange(newDPI);
+    }
+    // If the rich control is created on a screen with non-system DPI, an
+    // initial scaling is applied. This needs to be reset after the first DPI
+    // change. Reapply the styles to the text.
+    else if ( m_richBaseDPI != m_richUsedDPI )
+    {
+        wxEventBlocker block(this, wxEVT_TEXT);
+
+        wxSize baseDPI = m_richBaseDPI;
+
+        wxTextAttr attr;
+        wxTextAttr commonAttr;
+        int rangeStart = -1;
+        int rangeTotal = GetLastPosition();
+
+        for ( int i = 0; i <= rangeTotal; ++i )
+        {
+            // Get style requires the old baseDPI to determine the actually
+            // used font size
+            m_richBaseDPI = baseDPI;
+            GetStyle(i, attr);
+
+
+            if ( attr == commonAttr && i != rangeTotal )
+                continue;
+
+            if ( rangeStart >= 0 || i == rangeTotal )
```

Comment:
After staring at these conditions for 5 minutes, I still don't find them as obvious as I'd like. Could we write this loop like this instead:
```cpp
// Store the old base DPI value, it will be changed to the actually used
// new DPI below and will be left with this value after the end of this
// function, so we won't need to do anything during the next DPI change.
const wxSize baseDPI = m_richBaseDPI;

const int rangeTotal = GetLastPosition();
for ( int rangeStart = -1, i = 0; ; ++i )
{
    // Temporarily set base DPI to get the actually used font size.
    m_richBaseDPI = baseDPI;
    GetStyle(i, attr);
    m_richBaseDPI = m_richUsedDPI;

    if ( i == rangeTotal )
    {
        // Always set the attribute for the last span of text, we need
        // to do it if only to update its font size.
        SetStyle(rangeStart, i, commonAttr);
        break;
    }

    if ( attr == commonAttr )
    {
        // Still using the same attribute, don't change rangeStart below.
        continue;
    }

    // The attribute has changed, so apply the previous attribute to the
    // range it was used for, if non-empty.
    if ( rangeStart >= 0 )
    {
        SetStyle(rangeStart, i, commonAttr);
    }

    // And remember the new attribute, as well as the start of the range
    // it applies to.
    rangeStart = i;
    commonAttr = attr;
}
```

Personally I find this more clear, what do you think?

Reply:
I updated the loop to look like this.